rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -96,150 +105,327 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
+                            RecordSegmentReaderCreationError::InvalidNumberOfFiles => {
+                                error!("Error creating Record Segment: {:?}", e);
+                                return Err(
+                                    MergeMetadataResultsOperatorError::RecordSegmentCreationError(
+                                        *e,
+                                    ),
+                                );
+                            }
+                        }
                     }
                 };
+                for merged_id in merged_ids {
+                    // Skip already taken from log.
+                    if visited_ids.contains(&merged_id) {
+                        continue;
+                    }
+                    let record = match record_segment_reader
+                        .get_data_for_offset_id(merged_id)
+                        .await
+                    {
+                        Ok(record) => record,
+                        Err(e) => {
+                            tracing::error!("Error reading Record Segment: {:?}", e);
+                            return Err(MergeMetadataResultsOperatorError::RecordSegmentReadError);
+                        }
+                    };
 
-                ids.push(user_id.to_string());
-                metadata.push(record.metadata.clone());
-                match record.document {
-                    Some(document) => documents.push(Some(document.to_string())),
-                    None => documents.push(None),
+                    let user_id = match record_segment_reader
+                        .get_user_id_for_offset_id(merged_id)
+                        .await
+                    {
+                        Ok(user_id) => user_id,
+                        Err(e) => {
+                            println!("Error reading Record Segment: {:?}", e);
+                            return Err(MergeMetadataResultsOperatorError::RecordSegmentReadError);
+                        }
+                    };
+
+                    ids.push(user_id.to_string());
+                    metadata.push(record.metadata.clone());
+                    match record.document {
+                        Some(document) => documents.push(Some(document.to_string())),
+                        None => documents.push(None),
+                    }
                 }
             }
-        }
-
-        // Hydrate the data from the record segment for the remaining data
-        // if query_ids were specified
-        if input.remaining_query_ids.is_some() {
-            for query_id in input.remaining_query_ids.as_ref().unwrap().iter() {
-                let offset_id = match record_segment_reader
-                    .get_offset_id_for_user_id(query_id)
-                    .await
+            // Full scan.
+            None => {
+                let mut ids_in_log = HashSet::new();
+                // Log record.
+                for (log, _) in mat_records.iter() {
```

Comment:
This code feels quite redundant? Can we make it easier to read / more compact by coalescing?